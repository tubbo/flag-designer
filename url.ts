/**
 * Parameters that control flag generation.
 */
export interface FlagParams {
  /**
   * Type of division on the flag.
   */
  division: number

  /**
   * Shape that will be overlaid onto the flag.
   */
  overlay: number

  /**
   * Symbol on the top left corner of the flag.
   */
  symbol: number

  /**
   * Colors for the divisions of the flag.
   */
  divisionColors: number[]

  /**
   * Color of the overlay shape.
   */
  overlayColor: number

  /**
   * Color of the symbol.
   */
  symbolColor: number
}

function randomNumber(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1)) + min
}

function convertOptionsToParams(options: FlagParams) {
  return {
    d: options.division.toString(),
    c1: options.divisionColors[0].toString(),
    c2: options.divisionColors[1].toString(),
    c3: options.divisionColors[2].toString(),
    o: options.overlay.toString(),
    c4: options.overlayColor.toString(),
    s: options.symbol.toString(),
    c5: options.symbolColor.toString(),
  }
}

/**
 * Generate the URL to a flag SVG from a given set of parameters.
 */
export function generateFlagURL(options?: Partial<FlagParams>) {
  const url = new URL('https://flag-designer.appspot.com')
  const defaults: FlagParams = {
    division: randomNumber(1, 12),
    overlay: randomNumber(1, 12),
    symbol: randomNumber(1, 24),

    symbolColor: randomNumber(1, 8),
    overlayColor: randomNumber(1, 8),
    divisionColors: [
      randomNumber(1, 8),
      randomNumber(1, 8),
      randomNumber(1, 8),
    ],
  }
  const params = new URLSearchParams(
    convertOptionsToParams({ ...defaults, ...options })
  )

  url.pathname = '/gwtflags/SvgFileService'
  url.search = params.toString()

  return url.toString()
}
