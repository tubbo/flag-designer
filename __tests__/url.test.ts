import { generateFlagURL } from '../url.js'

describe('generateFlagURL', () => {
  test('generates url for an svg flag', () => {
    const url = generateFlagURL({
      division: 1,
      divisionColors: [2, 3, 4],
      overlay: 5,
      overlayColor: 6,
      symbol: 7,
      symbolColor: 8,
    })

    expect(url).toMatchSnapshot()
  })

  test('randomizes parameters when none given', () => {
    const url = generateFlagURL()

    expect(url).toStartWith('https://flag-designer.appspot.com')
  })

  test('randomizes parameters when partially given', () => {
    const url = generateFlagURL({
      division: 1,
    })

    expect(url).toMatch(/d=1/)
  })
})
