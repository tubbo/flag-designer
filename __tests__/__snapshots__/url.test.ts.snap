// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateFlagURL generates url for an svg flag 1`] = `"https://flag-designer.appspot.com/gwtflags/SvgFileService?d=1&c1=2&c2=3&c3=4&o=5&c4=6&s=7&c5=8"`;
